File: 2024/March/C/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MAX 10000

void quicksort(int data[], int tam);
void _qsort(int data[], int start, int end);
int partition(int data[], int start, int end);
void swap(int data[], int p1, int p2);

void quicksort(int data[], int tam)
{
    _qsort(data, 0, tam-1);
}

void _qsort(int data[], int start, int end) {
    if(start < end) {
        int pivot_pos = partition(data, start, end);
        _qsort(data, start, pivot_pos-1);
        _qsort(data, pivot_pos+1, end);
    }
}

int partition(int data[], int start, int end) {
    int i = start+1;
    int pivot = data[start];
    for(int j=start+1; j<=end; j++) {
        if(data[j] < pivot) {
            swap(data,i,j);
            i++;
        }
    }
    swap(data,start,i-1);
    return i-1;
}

void swap(int data[], int p1, int p2) {
    int tmp = data[p1];
    data[p1] = data[p2];
    data[p2] = tmp;
}

int main()
{
    int data[MAX];

    for(int i=0; i<MAX; i++)
        data[i] = rand()%(MAX*10);

    long start = clock();
    quicksort(data, MAX);
    long end = clock();

    //for(int i=0; i<MAX; i++)
    //    printf("%5d", data[i]);
    //printf("\n");
    printf("Tempo para %d elementos: %ld ns\n", MAX, (end-start));
}

File: 2024/March/Quicksort.java
package March;
import java.util.Random;

public class QuickSort {
    public static long quickSort(int[] arr) {
        return quickSort(arr, 0, arr.length - 1);
    }

    private static long quickSort(int[] arr, int low, int high) {
        long operationCount = 0;
        if (low < high) {
            int pi = partition(arr, low, high);
            operationCount += quickSort(arr, low, pi - 1);
            operationCount += quickSort(arr, pi + 1, high);
        }
        return operationCount;
    }

    private static int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = low - 1;
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        swap(arr, i + 1, high);
        return i + 1;
    }

    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static void main(String[] args) {
        Random r = new Random();
        int MAX = 10000;
        int[] data = new int[MAX];
        for (int i = 0; i < data.length; i++)
            data[i] = r.nextInt(data.length * 10);

        long operationCount = quickSort(data.clone());
        System.out.println("Número de operações do Quick Sort: " + operationCount);
    }
}